File: Android/Android-AV/app/src/main/java/com/dongnao/player/MainActivity.java
package com.dongnao.player;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private SystemMediaSelector mSystemMediaSelector;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1001);
        }

        System.out.println(DNPlayer.class);

        mSystemMediaSelector = new SystemMediaSelector(path -> startActivity(PlayActivity.newIntent(MainActivity.this, path)), this);
        findViewById(R.id.btnSelectFile).setOnClickListener(v -> mSystemMediaSelector.takeFile());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mSystemMediaSelector.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1001 && grantResults.length == 0) {
            supportFinishAfterTransition();
        }
    }

}

File: Android/Android-CV/app/src/main/java/me/ztiany/opencv/idrecognizing/MainActivity.kt
package me.ztiany.opencv.idrecognizing

import android.annotation.TargetApi
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.googlecode.tesseract.android.TessBaseAPI
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val log = fun(message: String) {
        Log.d("MainActivity", message)
    }

    private var tessBaseAPI: TessBaseAPI? = null

    private val systemMediaSelector by lazy {
        SystemMediaSelector(SystemMediaSelectorResultListener { path: String? ->
            log("path = $path")
            path?.let(::processImage)
        }, this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        askPermission()

        btnSelectImage.setOnClickListener {
            systemMediaSelector.takeFile("image/*")
        }

        btnInitTess.setOnClickListener {
            initTessTwo()
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun askPermission() {
        if (ContextCompat.checkSelfPermission(
                this,
                android.Manifest.permission.WRITE_EXTERNAL_STORAGE
            ) == PackageManager.PERMISSION_DENIED
        ) {
            requestPermissions(
                arrayOf(android.Manifest.permission.WRITE_EXTERNAL_STORAGE),
                PERMISSION_CODE
            )
        } else {
            initTessTwo()
        }
    }

    private fun initTessTwo() {
        TessTwoUtils.initTess(this, object : TessTwoUtils.CallBack {
            override fun onSuccess(result: TessBaseAPI?) {
                tessBaseAPI = result
                tvTessStatus.text = "Tess初始化成功"
            }

            override fun onProgressing() {
                Toast.makeText(this@MainActivity, "Tess is initializing", Toast.LENGTH_LONG).show()
                tvTessStatus.text = "Tess初始化中"
            }

            override fun onError() {
                Toast.makeText(this@MainActivity, "Tess init failed", Toast.LENGTH_LONG).show()
                tvTessStatus.text = "Tess初始化失败"
            }
        })
    }

    private fun processImage(path: String) {
        val bitmap = BitmapFactory.decodeFile(path)
        val number = ImageProcess.getIdNumber(BitmapUtils.toBitmap(path), Bitmap.Config.ARGB_8888)
        ivOrigin.setImageBitmap(bitmap)
        ivResult.setImageBitmap(number)

        val baseAPI = tessBaseAPI ?: return
        baseAPI.setImage(number)
        tvResult.text = baseAPI.utF8Text
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        systemMediaSelector.onActivityResult(requestCode, resultCode, data)
    }

    override fun onDestroy() {
        super.onDestroy()
        tessBaseAPI?.end()
    }

    companion object {
        private const val PERMISSION_CODE = 100
    }

}
